File: algorithms/2_medium/M/01594_maximum_non_negative_product_in_a_matrix/dp.c
#define MOD 1000000007

struct pair {
    int64_t max;
    int64_t min;
};

int maxProductPath(int *restrict *restrict grid, int rows, const int *restrict cols_data) {
    assert(rows > 0 && cols_data != NULL && cols_data[0] > 0 && grid != NULL);
    const int cols = cols_data[0];
    struct pair *buff = malloc(2 * (size_t)cols * sizeof(*buff));
    assert(buff != NULL);
    struct pair *restrict prev = &buff[0];
    struct pair *restrict curr = &buff[cols];

    prev[0] = (struct pair){ .max = grid[0][0], .min = grid[0][0] };
    for (int j = 1; j < cols; j += 1) {
        const int64_t v = prev[j - 1].min * grid[0][j];
        prev[j] = (struct pair){ .max = v, .min = v };
    }

    for (int i = 1; i < rows; i += 1) {
        const int64_t v = prev[0].min * grid[i][0];
        curr[0] = (struct pair){ .max = v, .min = v };

        for (int j = 1; j < cols; j += 1) {
            const int64_t cell = grid[i][j];
            const int64_t a = prev[j].min * cell;
            const int64_t b = prev[j].max * cell;
            const int64_t c = curr[j - 1].min * cell;
            const int64_t d = curr[j - 1].max * cell;

            curr[j] =  (struct pair){ .max = MAX(MAX(a, b), MAX(c, d)), .min = MIN(MIN(a, b), MIN(c, d)) };
        }
        struct pair *restrict temp = prev;
        prev = curr;
        curr = temp;
    }

    int res = 0;
    if (prev[cols - 1].max < 0) {
        res = -1;
    } else {
        res = (int)(prev[cols - 1].max % MOD);
    }

    free(buff);
    return res;
}

File: algorithms/2_medium/M/01594_maximum_non_negative_product_in_a_matrix/dp.rs
impl Solution {
    const MOD: i64 = 1_000_000_007;

    pub fn max_product_path(grid: Vec<Vec<i32>>) -> i32 {
        assert!(!grid.is_empty() && !grid[0].is_empty());
        let (rows, cols) = (grid.len(), grid[0].len());
        let mut buff = vec![(0_i64, 0_i64); 2 * cols];
        let (mut prev, mut curr) = buff.split_at_mut(cols);

        prev[0] = (grid[0][0] as i64, grid[0][0] as i64);
        for j in 1..cols {
            let v = (prev[j - 1].0 * grid[0][j] as i64);
            prev[j] = (v, v);
        }

        for i in 1..rows {
            let v = (prev[0].0 * grid[i][0] as i64);
            curr[0] = (v, v);

            for j in 1..cols {
                let cell = grid[i][j] as i64;
                let a = (prev[j].0 * cell);
                let b = (prev[j].1 * cell);
                let c = (curr[j - 1].0 * cell);
                let d = (curr[j - 1].1 * cell);

                curr[j] = (a.min(b).min(c).min(d), a.max(b).max(c).max(d));
            }
            (curr, prev) = (prev, curr);
        }

        if prev[cols - 1].1 < 0 {
            return -1;
        }

        (prev[cols - 1].1 % Self::MOD) as i32
    }
}
